### ISSUE TITLE
Missing horizontal scroll motions in vim mode

### ISSUE THREAD

**Issue by TrevorHinesley** [zed-industries/zed#17219]
### Check for existing issues

- [X] Completed

### Describe the feature

Example from vim's `help :scroll-horizontal`:

```
zL          Move the view on the text half a screenwidth to the
            right, thus scroll the text half a screenwidth to the
            left.  This only works when 'wrap' is off.  {not in
            Vi}


zH          Move the view on the text half a screenwidth to the
            left, thus scroll the text half a screenwidth to the
            right.  This only works when 'wrap' is off.  {not in
            Vi}
```

**Comment by dinocosta** [comment]
Hey there @ConradIrwin 👋

I'm taking a look at this issue and was wondering if there's any method that allows a caller to determine the editor's width, or something similar to `editor::scroll::ScrollManager.visible_line_count`, but for the number of columns. 

I have an initial implementation (in this [branch](https://github.com/dinocosta/zed/tree/17219-vim-horizontal-scroll)) but it's simply updating the scroll's manager scroll anchor's offset. Although this does scroll the editor horizontally, it doesn't exactly behave as vim as we'd need to:

1. Scroll half of the editor's width
2. Move the cursor to the scrolled position

The second point should be achievable with vim's `move_cursor` method, but I'm not so sure about the first point. Let me know if you have any suggestions, thanks!

P.S.: I've also booked a pairing session, in case you'd rather discuss this during the call.

**Comment by ConradIrwin** [comment]
@dinocosta I don't think we have that information yet. Last time I looked into this with a collaborator, we found that there was some confusion between `em_width` and `em_advance` in terms of setting the scroll position (though that was for `zh` and `zl` not this).

The editor has a `last_position_map` with a `text_hitbox` which probably can be used to calculate "half a screen width" (though you may need to play around to see if it matches). Once you have the position you want to be at, you can get back to a character offset by using the PositionMap's `point_for_position`.

**Comment by dinocosta** [comment]
> The editor has a `last_position_map` with a `text_hitbox` which probably can be used to calculate "half a screen width" (though you may need to play around to see if it matches). Once you have the position you want to be at, you can get back to a character offset by using the PositionMap's `point_for_position`.

Thanks! This has been very helpful already, I've managed to update my implementation to leverage the editor's `last_position_map.text_hitbox.bouns.size.width` and halving that value does seems to correctly scroll the editor half of the screenwidth.

I'll try to now explore updating the implementation to also leverage `point_for_position` and use a `DisplayPoint` instead, for which we can then obtain a new anchor, like it's being done in `set_scroll_position`. Hopefully this will also make it easier to not only scroll the editor but also move the cursor to that new point/anchor 🙌

**Comment by alex35mil** [comment]
It would be great to be able to configure amount of lines to be scrolled via action argument: https://github.com/zed-industries/zed/discussions/32290

**Comment by ConradIrwin** [comment]
@alex35mil I think you should be able to do this with the keymap, so no need for a different setting:

```
[
{
  "context": "Editor && !VimCount",
  "z shift-l": ["workspace::SendKeystrokes": "5 z shift-l"]
}
]
```

**Comment by alex35mil** [comment]
@ConradIrwin It doesn’t work for insert mode.

**Comment by ConradIrwin** [comment]
Oops! You'll also want `VimControl` to limit it to normal/visual mode

**Comment by alex35mil** [comment]
But I need the scroll in insert mode, too. When I type something and realize I need a bit more context, I do a scrolling in vim without interrupting insert mode. It’s not possible in Zed.

**Comment by ConradIrwin** [comment]
How did you set that up in vim? In theory zed should work the same way (though we don't currently show a preview of the typed characters if you do a `z`, that should ship to preview on Wednesday)

**Comment by alex35mil** [comment]
I [use `normal`](https://github.com/alex35mil/dotfiles/blob/2d3db3edba60093199c5523325a33601e80e2556/home/.config/nvim/lua/editor/navigation.lua#L23-L24) in Vim.

**Comment by ConradIrwin** [comment]
Ah ok. Trickier :D.

We do want to implement `:normal`. In the meantime, you should be able to use `ctrl-o` to get to temporary normal mode (though I haven't tried it).


[
  {
    "context": "Editor && !VimCount && vim_mode == insert",
    "z shift-l": ["workspace::SendKeystrokes": "ctrl-o 5 z shift-l"],
  }
]

**Comment by alex35mil** [comment]
I tried this:

```json
{
    "context": "Editor",
    "bindings": {
        "cmd-shift-up": "vim::LineUp",
        "cmd-shift-down": "vim::LineDown",
        "ctrl-shift-up": ["workspace::SendKeystrokes", "ctrl-o 5 cmd-shift-up"],
        "ctrl-shift-down": ["workspace::SendKeystrokes", "ctrl-o 5 cmd-shift-down"],
    }
}
```

The result is odd.

In insert mode, when I hit `ctrl-shift-down`, it scrolls 5 down, on the next hit, 1 down, then 5 down again, then 1, and so on.

In normal mode, it does different things. Most of the time, it just scrolls 1 line up/down. But sometimes, I don't even understand what it does - on the first hit, it jumps to some random location, then on the next hit it inserts `5`, then on the next one, scrolls one line up/down. I can't reliably reproduce it, though.

**Comment by ConradIrwin** [comment]
Sorry about that, a confusing set of bugs!

In normal mode, `ctrl-o` jumps to the "last position in the jump list" (which until we merge https://github.com/zed-industries/zed/pull/30345 contains recent scroll positions).

In insert mode, it looks like we don't exit temporary normal mode after scroll events (fix https://github.com/zed-industries/zed/pull/32653).

So I think I need to get both these PRs merged, and you need to have a different binding in normal mode (with no ctrl-o) vs insert mode (with ctrl-o).

(And/or we ship `:norm`...)
